GAE/src/org/waterforpeople/mapping/dataexport/GraphicalSurveySummaryExporter.java

```diff
@@ -529,80 +532,96 @@ private void analyseData(final List<InstanceData> allData) {
         }
     }
 
+    /**
+     * @param sheet
+     * @param instanceData
+     * @param whichQuestions
+     * @param md
+     * @param groupSheet
+     * @throws NoSuchAlgorithmException
+     */
     private synchronized void writeInstanceDataSplit(
             Sheet sheet,
             InstanceData instanceData,
             List<QuestionDto> whichQuestions,
-            List<Row> digestRows,
-            boolean showRepeatColumn)
+            MessageDigest md,
+            boolean groupSheet)
             throws NoSuchAlgorithmException {
 
         int startRow = sheet.getLastRowNum() + 1;
-
-        for (QuestionDto questionDto : whichQuestions) {
-            final Long questionId = questionDto.getKeyId();
-
+        //TODO maxIterationsCount is for entire instance. We need it for this sheet only!
+        int maxIter = 0;
+        if (!groupSheet) { //Must have one row of metadata on base sheet, even if no values
+            maxIter = 1;
+        }
+        for (QuestionDto qd : whichQuestions) {
+            final Long questionId = qd.getKeyId();
             SortedMap<Long, String> iterationsMap = instanceData.responseMap.get(questionId);
-            if (iterationsMap == null) {
-                continue;
+            if (iterationsMap != null) {
+                maxIter = Math.max(maxIter, iterationsMap.size());
             }
+        }
 
-            // Write downwards (and possibly rightwards) per iteration
-            int currentRow = startRow - 1;
-            for (Map.Entry<Long, String> iteration : iterationsMap.entrySet()) {
-                String val = iteration.getValue();
-                Row iterationRow = getRow(++currentRow, sheet);
-                writeAnswer(iterationRow, columnIndexMap.get(questionId.toString()),
-                        questionDto, val);
-                if (!digestRows.contains(iterationRow)) { //A Set would be neater
-                    digestRows.add(iterationRow);
+        for (int i = 0; i < maxIter; i++) {
+            Row iterationRow = getRow(startRow + i, sheet);
+            writeMetadataRow(iterationRow, instanceData, i + 1, groupSheet);
+            for (QuestionDto qd : whichQuestions) {
+                final Long questionId = qd.getKeyId();
+                final QuestionDto questionDto = questionsById.get(questionId);
+                SortedMap<Long, String> iterationsMap = instanceData.responseMap.get(questionId);
+                if (iterationsMap == null) {
+                    continue; //No answers on this sheet
+                }
+                String val = iterationsMap.get(new Long(i));
```

Comment:
It feels weird to have a `new Long()` here.  Maybe its worth it converting that and related code to an `int`?

Reply:
Then we cannot use a Map, it depends on the mapping being Object->Object